app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,229 @@
+        showToast(activity,
+            activity.getString(R.string.discovery_failed),
+            Toast.LENGTH_SHORT);
+      }
+    });
+  }
+
+  public void setWifiP2pEnabled(boolean wifiP2pEnabled) {
+    this.wifiP2pEnabled = wifiP2pEnabled;
+  }
+
+  public boolean isWifiP2pEnabled() {
+    return wifiP2pEnabled;
+  }
+
+  /* From WifiP2pManager.ChannelListener interface */
+  @Override
+  public void onChannelDisconnected() {
+    // Upon disconnection, retry one more time
+    if (manager != null && !retryChannel) {
+      Log.d(TAG, "Channel lost, trying again");
+      activity.clearPeers();
+      retryChannel = true;
+      manager.initialize(activity, getMainLooper(), this);
+    } else {
+      showToast(activity, R.string.severe_loss_error, Toast.LENGTH_LONG);
+    }
+  }
+
+  /* From WifiP2pManager.PeerListListener callback-interface */
+  @Override
+  public void onPeersAvailable(@NonNull WifiP2pDeviceList peers) {
+    ((Callbacks) activity).updatePeerDevicesList(peers);
+  }
+
+  /* From WifiP2pManager.ConnectionInfoListener callback-interface */
+  @Override
+  public void onConnectionInfoAvailable(@NonNull WifiP2pInfo groupInfo) {
+    /* Devices have successfully connected, and 'info' holds information about the wifi p2p group formed */
+    setGroupInfo(groupInfo);
+    ((Callbacks) activity).performHandshakeWithSelectedPeerDevice(groupInfo);
+  }
+
+  public void setUserDevice(@NonNull WifiP2pDevice userDevice) {
+    this.userDevice = userDevice;
+  }
+
+  public void setGroupInfo(@NonNull WifiP2pInfo groupInfo) {
+    this.groupInfo = groupInfo;
+  }
+
+  public boolean isGroupFormed() {
+    return groupInfo.groupFormed;
+  }
+
+  public boolean isGroupOwner() {
+    return groupInfo.isGroupOwner;
+  }
+
+  public @NonNull InetAddress getGroupOwnerAddress() {
+    return groupInfo.groupOwnerAddress;
+  }
+
+  public void setSenderSelectedPeerDevice(@NonNull WifiP2pDevice senderSelectedPeerDevice) {
+    this.senderSelectedPeerDevice = senderSelectedPeerDevice;
+  }
+
+  public void connect() {
+    if(senderSelectedPeerDevice == null) {
+      Log.d(TAG, "No device set as selected");
+    }
+
+    WifiP2pConfig config = new WifiP2pConfig();
+    config.deviceAddress = senderSelectedPeerDevice.deviceAddress;
+    config.wps.setup = WpsInfo.PBC;
+
+    manager.connect(channel, config, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        // UI updated from broadcast receiver
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.connection_failed) + ": " + errorMessage);
+        showToast(activity, activity.getString(R.string.connection_failed),
+            Toast.LENGTH_LONG);
+      }
+    });
+  }
+
+  public void closeLocalFileTransferActivity() {
+    activity.cancelAsyncTasks();
```

Comment:
you already call cancelAsynctasks in onDestroy and when you call `finish` then `onDestroy` is called.

Regardless the asynctasks should probably live in WifiDirectManager and not the activity

Reply:
> you already call cancelAsynctasks in onDestroy and when you call finish then onDestroy is called.

In case of the sender device, the senderDeviceAsyncTask directly calls `finish()` on the activity, it doesn't call `closeLocalFileTransferActivity()`

That is because `closeLocalFileTransferActivity()` also disconnects the wifi-direct connection, because of which the last file being transferred always gets corrupted (what I could make out was that there's a lag between the sender finishing adding the file to the output stream, and the receiver reading that stream completely. Disconnecting during that lag causes loss of data, hence the file corruption.)

> Regardless the asynctasks should probably live in WifiDirectManager and not the activity

Could you explain the benefit of this? (I can't figure out the advantage of placing it in either of the two)

Reply:
LocalFileTransferActivity should not know the implementation of WifiDirectManager and this is definitely a impl detail.
Once you get to the task in the future of rewriting these in rx you will only have to change WifiDirectManger and not your client code and that will be much cleaner and efficient in and of itself. It doesn't really make sense for LocalFileTransferActivity to change because we altered how we do things asynchronously does it? It just renders data, it can't care about how that data was gotten.
We are getting ever closer to a portable feature completely contained to itself, we could nearly reuse this WifiDirectManager in any activity with only a few lines and that is a mark of a nice component.
It is also getting more and more testable for whenever you wish to make a foray into that.

Reply:
`last file being transferred always gets corrupted` if this is the case then introduce a delay in SenderTask so that the transfer actually completes when that task completes, sounds like a very odd bug

Reply:
With this [commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/d11af7ba65fa96b4f7f5bac5e69a985cf06c54f3), I think this particular thread has been resolved:
1) There's no closeLocalFileTransferActivity() anymore, just simple finish() calls. The cleanup is handled by `destroyWifiDirectManager()` - [commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/19b316644e88a12281b92b890d156bea07572ddd)

2) The file transfer code and async tasks have been shifted to WDM - [commit](https://github.com/kiwix/kiwix-android/pull/1276/commits/d11af7ba65fa96b4f7f5bac5e69a985cf06c54f3)

3) The issue of last file getting corrupted is side-stepped by having only the receiver disconnect. (For cleanup only require releasing the channel, which is done by both).